text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,667 @@
+*must* be marked `unsafe` because it delegates the responsibility of maintaining `Vec`'s safety
+invariants  to its callers. However, the implementation of [`Vec::set_len`] does not contain any
+explicitly `unsafe` operations. Consequently, there is no tooling cue that suggests this function
+should be unsafe — doing so is entirely a matter of programmer discipline.
+
+Providing tooling support for field safety hygiene will close this gap in the tooling for function
+safety hygiene.
+
+### Benefit: Making Unsafe Rust Easier to Audit
+
+As a consequence of improving function and field safety hygiene, the process of auditing internally
+`unsafe` abstractions will be made easier in at least two ways. First, as previously discussed, we
+anticipate that tooling support for field safety hygiene will encourage programmers to document
+when their fields carry safety invariants.
+
+Second, we anticipate that good field safety hygiene will narrow the scope of safety audits.
+Presently, to evaluate the soundness of an `unsafe` block, it is not enough for reviewers to *only*
+examine `unsafe` code; the invariants upon which `unsafe` code depends may also be violated in safe
+code. If `unsafe` code depends upon field safety invariants, those invariants may presently be
+violated in any safe (or unsafe) context in which those fields are visible. So long as Rust permits
+safety invariants to be violated at-a-distance in safe code, audits of unsafe code must necessarily
+consider distant safe code. (See [*The Scope of Unsafe*].)
+
+[*The Scope of Unsafe*]: https://www.ralfj.de/blog/2016/01/09/the-scope-of-unsafe.html
+
+For crates that practice good safety hygiene, reviewers will mostly be able to limit their review
+of distant routines to only `unsafe` code.
+
+# Guide-level explanation
+
+A safety invariant is any boolean statement about the computer at a time *t*, which should remain
+true or else undefined behavior may arise. Language safety invariants are imposed by the Rust
+itself and must never be violated; e.g., a `NonZeroU8` must *never* be 0.
+
+Library safety invariants, by contrast, are imposed by an API. For example, `str` encapsulates
+valid UTF-8 bytes, and much of its API assumes this to be true. This invariant may be temporarily
+violated, so long as no code that assumes this safety invariant holds is invoked.
+
+Safety hygiene is the practice of denoting and documenting where memory safety obligations arise
+and where they are discharged. To denote that a field carries a library safety invariant, use the
+`unsafe` keyword in its declaration and document its invariant; e.g.:
+
+```rust
+pub struct UnalignedRef<'a, T> {
+    /// # Safety
+    /// 
+    /// `ptr` is a shared reference to a valid-but-unaligned instance of `T`.
+    unsafe ptr: *const T,
+    _lifetime: PhantomData<&'a T>,
+}
+```
+
+(Note, the `unsafe` field modifier is only applicable to named fields. You should avoid attaching
+library safety invariants to unnamed fields.)
```

Comment:
This would not be too hard to add, I think. Parentheses could be used to disambiguate:

```rust
// safe field of type `unsafe fn()`
struct Foo(unsafe fn());

// unsafe field of type `fn()`
struct Foo(unsafe (fn()));
```